near-lake-framework/src/streamer.rs

```diff
@@ -0,0 +1,230 @@
+            tracing::debug!(target: crate::LAKE_FRAMEWORK, "Fetching a list of blocks from S3...");
+            match crate::s3_fetchers::list_blocks(
+                s3_client,
+                s3_bucket_name,
+                start_from_block_height,
+                number_of_blocks_requested
+            )
+            .await {
+                Ok(block_heights) => {
+                    if block_heights.is_empty() {
+                        tracing::debug!(
+                            target: crate::LAKE_FRAMEWORK,
+                            "There are no newer block heights than {} in bucket {}. Fetching again in 2s...",
+                            start_from_block_height,
+                            s3_bucket_name,
+                        );
+                        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
+                        continue;
+                    }
+                    tracing::debug!(
+                        target: crate::LAKE_FRAMEWORK,
+                        "Received {} newer block heights",
+                        block_heights.len()
+                    );
+
+                    start_from_block_height = *block_heights.last().unwrap() + 1;
+                    for block_height in block_heights {
+                        tracing::debug!(target: crate::LAKE_FRAMEWORK, "Yielding {} block height...", block_height);
+                        yield block_height;
+                    }
+                }
+                Err(err) => {
+                    tracing::warn!(
+                        target: crate::LAKE_FRAMEWORK,
+                        "Failed to get block heights from bucket {}: {}. Retrying in 1s...",
+                        s3_bucket_name,
+                        err,
+                    );
+                    tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+                }
+            }
+        }
+    }
+}
+
+async fn fast_fetch_block_heights(
+    pending_block_heights: &mut std::pin::Pin<&mut impl tokio_stream::Stream<Item = u64>>,
+    limit: usize,
+    await_for_at_least_one: bool,
+) -> anyhow::Result<Vec<u64>> {
+    let mut block_heights = Vec::with_capacity(limit);
+    for remaining_limit in (0..limit).rev() {
+        tracing::debug!(target: crate::LAKE_FRAMEWORK, "Polling for the next block height without awaiting... (up to {} block heights are going to be fetched)", remaining_limit);
+        match futures::poll!(pending_block_heights.next()) {
+            std::task::Poll::Ready(Some(block_height)) => {
+                block_heights.push(block_height);
+            }
+            std::task::Poll::Pending => {
+                if await_for_at_least_one && block_heights.is_empty() {
+                    tracing::debug!(target: crate::LAKE_FRAMEWORK, "There were no block heights available immediately, and the prefetching blocks queue is empty, so we need to await for at least a single block height to be available before proceeding...");
```

Comment:
I enjoy reading your messages, really, it's awesome

Reply:
Redirecting this particular kudos to @frol